terraform/files/bin/update-R4-to-R5.sh

```diff
@@ -0,0 +1,106 @@
+#
+# This script updates cluster files cluster-template.yaml and clusterctl.yaml from R4 to R5 version of these files.
+# To tweak cluster files of existing cluster use script as follows:
+#    ```sh
+#    ./update-R4-to-R5.sh <CLUSTER_NAME>
+#    ```
+# Alternatively, you can use `cluster-defaults` to change the cluster templates in `~/cluster-defaults/` which get used
+# when creating new clusters as follows:
+#    ```sh
+#    ./update-R4-to-R5.sh cluster-defaults
+#    ```
+#
+# (c) Matej Feder <feder.mato@gmail.com>, 7/2023
+# SPDX-License-Identifier: Apache-2.0
+
+
+usage()
+{
+  echo "!This script is under development and will be stabilized within the release R5!"
+  echo "Usage: update-R4-to-R5.sh <CLUSTER_NAME>"
+  echo "Updates cluster files from R4 to R5 version."
+  exit 1
+}
+
+restore()
+{
+  echo "Updating failed ($1)" 1>&2
+  cp -p cluster-template.yaml.backup cluster-template.yaml
+  cp -p clusterctl.yaml.backup clusterctl.yaml
+  exit "$1"
+}
+
+CLUSTER_NAME="$1"
+if test -z "$CLUSTER_NAME"; then
+  usage
+else
+  cd ~/"${CLUSTER_NAME}" || { echo "Cluster config $CLUSTER_NAME does not exist" 1>&2; exit 2; }
+fi
+
+# Backup files
+cp -p cluster-template.yaml cluster-template.yaml.backup
+cp -p clusterctl.yaml clusterctl.yaml.backup
+
+# Update cluster-template.yaml
+# TODO: Update `update-cluster-template.diff` file when R5 release will be stabilized
+if grep -q "SERVICE_CIDR\|POD_CIDR\|# Defragment & backup & trim script for SCS k8s-cluster-api-provider etcd cluster.\|# Allow to configure registry hosts in containerd" cluster-template.yaml; then
+  echo "cluster-template.yaml already updated"
+else
+  # The default template file `cluster-defaults/cluster-template.yaml` of version R4 still references the "old" `k8s.gcr.io` container registry.
+  # This container registry string was patched by the `fixup_k8sregistry.sh` script within `create_cluster.sh`.
+  # Therefore, the R5 patch for the default template file `cluster-defaults/cluster-template.yaml` and for
+  # workload cluster template file `<CLUSTER_NAME>/cluster-template.yaml` is different.
+  if test "$CLUSTER_NAME" != "cluster-defaults"; then
+    patch cluster-template.yaml < ~/k8s-cluster-api-provider/terraform/files/update/R4_to_R5/update-cluster-template.diff || restore 3
+  else
+    patch cluster-template.yaml < ~/k8s-cluster-api-provider/terraform/files/update/R4_to_R5/update-cluster-template-cluster-defaults.diff || restore 4
+  fi
+  # Ensure that the directory for containerd registry configs exists and create a default containerd registry
+  # config file that instructs containerd to use registry.scs.community container registry
+  # instance as a public mirror of DockerHub. See #432 for further details.
+  mkdir -p containerd/hosts
+  cat >containerd/hosts/docker.io <<EOF
+server = "https://registry-1.docker.io"
+[host."https://registry.scs.community/v2/docker.io"]
+    capabilities = ["pull"]
+    override_path = true
+EOF
```

Comment:
This is true. In the "expected" scenario this file could be copied from the updated (R5) k8s-cluster-api-provider clone in the management node.  
As this is only one dependency on the R5 source code so far, I decided to create it inside the script, just because of better script testability, i.e. in the current version you can test this script without the R5 k8s-cluster-api-provider clone on the right path.  I also expect that this file won't be changed until R5. But if there will be another similar case until R5 I will reconsider also this one.

Reply:
Thanks for the explanation

Reply:
I changed my mind, you are right, fixed as you proposed